lua/entities/gmod_tardis/modules/sh_chameleon.lua

```diff
@@ -0,0 +1,363 @@
+
+        local new_percent = math.Approach(percent, target, speed * timepassed) or 1
+        local high_percent = math.Clamp(new_percent + 0.5, 0, 1) or 1
+        self:SetData("chameleon-phase-percent", new_percent)
+        self:SetData("chameleon-phase-highPercent", high_percent)
+
+        local modelmaxs = self:GetData("modelmaxs")
+        local modelheight = self:GetData("modelheight")
+
+        local pos = self:GetPos() + self:GetUp() * (modelmaxs.z - (modelheight * high_percent))
+        local pos2 = self:GetPos() + self:GetUp() * (modelmaxs.z - (modelheight * new_percent))
+
+        self:SetData("phase-highPos", pos)
+        self:SetData("phase-pos", pos2)
+    end)
+
+    ENT:AddHook("ShouldDrawPhaseAnimation", "chameleon", function(self)
+        if self:GetData("chameleon_animation",false) then
+            return true
+        end
+    end)
+
+    ENT:AddHook("ExteriorChanged", "scale", function(self)
+        self:DisableMatrix("RenderMultiply")
+    end)
+end
+
+function ENT:ChangeExteriorMetadata(id)
+    if SERVER then
+        self:SendMessage("exterior_metadata_update", {id})
+    end
+
+    if self.metadata.ExteriorOriginal == nil then
+        self.metadata.ExteriorOriginal = self.metadata.Exterior
+    end
+    local original_md = self.metadata.ExteriorOriginal
+
+    local ext_md = (id == false and original_md) or TARDIS:CreateExteriorMetadata(id)
+
+    local oldvortex = self.metadata.Exterior.Parts.vortex
+    if oldvortex then
+        ext_md.Parts.vortex = TARDIS:CopyTable(oldvortex)
+    end
+
+    ext_md.Teleport = TARDIS:CopyTable(original_md.Teleport)
+    ext_md.Sounds.Teleport = TARDIS:CopyTable(original_md.Sounds.Teleport)
+
+    if original_md.ProjectedLight and original_md.ProjectedLight.color then
+        ext_md.ProjectedLight = ext_md.ProjectedLight or {}
+        ext_md.ProjectedLight.color = original_md.ProjectedLight.color
+    end
+
+    self.metadata.Exterior = ext_md
+    self.interior.metadata.Exterior = ext_md
+
+    return ext_md
+end
+
+if CLIENT then
+    ENT:OnMessage("exterior_metadata_update", function(self,data,ply)
+        self:ChangeExteriorMetadata(data[1])
+    end)
+end
+
+function ENT:ChangeExterior(id, animate, ply, retry)
+    if CLIENT then
+        self:SendMessage("chameleon_change_exterior", {id, animate, ply, retry})
+        return
+    end
+
+    if id == nil then
+        id = false
+    end
+
+    local can_apply,select_failed,msg,msg_is_err = self:CallCommonHook("CanChangeExterior", id, retry)
+
+    if can_apply == false then
+        if select_failed then
+            TARDIS:ErrorMessage(ply, "Chameleon.FailedExteriorSelect")
+        else
+            self:SetData("chameleon_selected_exterior", id, true)
+            self:SetData("chameleon_exterior_last_selector", ply, true)
+            TARDIS:Message(ply, "Chameleon.ExteriorSelected")
+        end
+
+        if msg then
+            if msg_is_err then
+                TARDIS:ErrorMessage(ply, msg)
```

Comment:
Can we combine this with the previous error when it exists? It looks a bit ugly having two appear at the same time. You can accomplish this using string interpolation in the language stuff e.g. `%s` 

See `Interior.FindingPositionFailed.Generic` for an example

Reply:
Sometimes there are 2 errors, sometimes the second message is not an error

Reply:
There are several cases in which we send several notifications at once.
I would prefer not to make these too long 

> we should combine them with the informational / error message when they exist
We can combine two errors or two normal notifications (which wouldn't look good in my opinion)
But how do you propose we combine an error message with a normal notification message?
Set the type of both to one?

Reply:
Closing as discussed